File: src/parsers/python.ts
import type Parser from 'web-tree-sitter';
import { SymbolFlag, SymbolKind, type ExportedSymbol } from '../index/types';
import { unquote, type ExtractionResult, type ReExportEntry, type ReExportName } from './base';

type SN = Parser.SyntaxNode;

export function extractPython(tree: Parser.Tree, respectAll: boolean): ExtractionResult {
  const root = tree.rootNode;
  let hasAll = false;
  const allList: string[] = [];
  const candidates: ExportedSymbol[] = [];
  const reExports: ReExportEntry[] = [];

  for (const c of root.namedChildren) {
    if (c.type === 'expression_statement') {
      const inner = c.namedChild(0);
      if (inner?.type === 'assignment') {
        const left = inner.childForFieldName('left');
        if (left && left.type === 'identifier' && left.text === '__all__') {
          hasAll = true;
          collectAllItems(inner.childForFieldName('right'), allList);
          continue;
        }
        if (left && left.type === 'identifier') {
          candidates.push(mkSym(left.text, SymbolKind.Variable, c));
        }
      }
      continue;
    }

    let target = c;
    if (c.type === 'decorated_definition') {
      const def = c.childForFieldName('definition');
      if (def) target = def;
    }

    if (target.type === 'function_definition') {
      const name = target.childForFieldName('name')?.text;
      if (name) candidates.push(mkSym(name, SymbolKind.Function, target));
    } else if (target.type === 'class_definition') {
      const name = target.childForFieldName('name')?.text;
      if (name) candidates.push(mkSym(name, SymbolKind.Class, target));
    } else if (c.type === 'import_from_statement') {
      handleImportFrom(c, reExports);
    }
  }

  let exports: ExportedSymbol[];
  if (respectAll && hasAll) {
    const allow = new Set(allList);
    exports = candidates.filter((s) => allow.has(s.name));
  } else {
    exports = candidates.filter((s) => !s.name.startsWith('_'));
  }
  return { exports, reExports };
}

function handleImportFrom(node: SN, reOut: ReExportEntry[]): void {
  const moduleNode = node.childForFieldName('module_name');
  if (!moduleNode) return;
  const fromPath = moduleNode.text;

  const names: ReExportName[] = [];
  let sawWildcard = false;

  for (const c of node.children) {
    if (c === moduleNode) continue;
    if (c.type === 'wildcard_import' || c.text === '*') {
      sawWildcard = true;
    } else if (c.type === 'dotted_name' && c !== moduleNode) {
      names.push({ exportedName: c.text });
    } else if (c.type === 'aliased_import') {
      const nm = c.childForFieldName('name');
      const al = c.childForFieldName('alias');
      if (nm) {
        names.push({
          exportedName: al?.text ?? nm.text,
          sourceName: al ? nm.text : undefined,
        });
      }
    }
  }

  if (sawWildcard) {
    reOut.push({ fromPath, names: 'all', isTypeOnly: false });
  } else if (names.length > 0) {
    reOut.push({ fromPath, names, isTypeOnly: false });
  }
}

function collectAllItems(node: SN | null, out: string[]): void {
  if (!node) return;
  if (node.type === 'list' || node.type === 'tuple') {
    for (const item of node.namedChildren) {
      if (item.type === 'string') {
        out.push(stringLiteralValue(item));
      }
    }
  }
}

function stringLiteralValue(node: SN): string {
  for (const c of node.namedChildren) {
    if (c.type === 'string_content') return c.text;
  }
  return unquote(node.text);
}

function mkSym(name: string, kind: SymbolKind, node: SN): ExportedSymbol {
  return {
    name,
    kind,
    flags: SymbolFlag.None,
    line: node.startPosition.row,
    col: node.startPosition.column,
  };
}

File: daemon/src/parsers/python.rs
use tree_sitter::{Node, Tree};

use crate::index::{ExportedSymbol, SymbolFlag, SymbolKind};

use super::extractor::{unquote, ExtractionResult, ReExportEntry, ReExportName, ReExportNames};

pub fn extract_python(tree: &Tree, source: &str, respect_all: bool) -> ExtractionResult {
    let mut result = ExtractionResult::default();
    let root = tree.root_node();
    let bytes = source.as_bytes();

    let mut has_all = false;
    let mut all_list: Vec<String> = Vec::new();
    let mut candidates: Vec<ExportedSymbol> = Vec::new();

    let mut cursor = root.walk();
    for c in root.named_children(&mut cursor) {
        let kind = c.kind();
        if kind == "expression_statement" {
            if let Some(inner) = c.named_child(0) {
                if inner.kind() == "assignment" {
                    let left = inner.child_by_field_name("left");
                    if let Some(left) = left {
                        if left.kind() == "identifier" {
                            let text = left.utf8_text(bytes).unwrap_or("");
                            if text == "__all__" {
                                has_all = true;
                                if let Some(right) = inner.child_by_field_name("right") {
                                    collect_all_items(right, bytes, &mut all_list);
                                }
                                continue;
                            }
                            candidates.push(mk_sym(text, SymbolKind::Variable, &c));
                        }
                    }
                }
            }
            continue;
        }

        let target = if kind == "decorated_definition" {
            c.child_by_field_name("definition").unwrap_or(c)
        } else {
            c
        };

        match target.kind() {
            "function_definition" => {
                if let Some(name_node) = target.child_by_field_name("name") {
                    if let Ok(name) = name_node.utf8_text(bytes) {
                        candidates.push(mk_sym(name, SymbolKind::Function, &target));
                    }
                }
            }
            "class_definition" => {
                if let Some(name_node) = target.child_by_field_name("name") {
                    if let Ok(name) = name_node.utf8_text(bytes) {
                        candidates.push(mk_sym(name, SymbolKind::Class, &target));
                    }
                }
            }
            _ => {}
        }

        if kind == "import_from_statement" {
            handle_import_from(c, bytes, &mut result.re_exports);
        }
    }

    if respect_all && has_all {
        let allow: std::collections::HashSet<&String> = all_list.iter().collect();
        result.exports = candidates
            .into_iter()
            .filter(|s| allow.contains(&s.name))
            .collect();
    } else {
        result.exports = candidates
            .into_iter()
            .filter(|s| !s.name.starts_with('_'))
            .collect();
    }
    result
}

fn handle_import_from(node: Node, bytes: &[u8], re_exports: &mut Vec<ReExportEntry>) {
    let Some(module_node) = node.child_by_field_name("module_name") else {
        return;
    };
    let Ok(from_path) = module_node.utf8_text(bytes) else {
        return;
    };

    let mut saw_wildcard = false;
    let mut names: Vec<ReExportName> = Vec::new();

    let mut cursor = node.walk();
    for c in node.children(&mut cursor) {
        if c == module_node {
            continue;
        }
        match c.kind() {
            "wildcard_import" => {
                saw_wildcard = true;
            }
            "*" => {
                saw_wildcard = true;
            }
            "dotted_name" => {
                if let Ok(text) = c.utf8_text(bytes) {
                    names.push(ReExportName {
                        exported_name: text.to_string(),
                        source_name: None,
                    });
                }
            }
            "aliased_import" => {
                let name = c.child_by_field_name("name");
                let alias = c.child_by_field_name("alias");
                if let Some(name) = name {
                    let name_text = name.utf8_text(bytes).unwrap_or("");
                    if let Some(alias) = alias {
                        let alias_text = alias.utf8_text(bytes).unwrap_or("");
                        names.push(ReExportName {
                            exported_name: alias_text.to_string(),
                            source_name: Some(name_text.to_string()),
                        });
                    } else {
                        names.push(ReExportName {
                            exported_name: name_text.to_string(),
                            source_name: None,
                        });
                    }
                }
            }
            _ => {}
        }
    }

    if saw_wildcard {
        re_exports.push(ReExportEntry {
            from_path: from_path.to_string(),
            names: ReExportNames::All,
            is_type_only: false,
        });
    } else if !names.is_empty() {
        re_exports.push(ReExportEntry {
            from_path: from_path.to_string(),
            names: ReExportNames::Named(names),
            is_type_only: false,
        });
    }
}

fn collect_all_items(node: Node, bytes: &[u8], out: &mut Vec<String>) {
    if node.kind() != "list" && node.kind() != "tuple" {
        return;
    }
    let mut cursor = node.walk();
    for item in node.named_children(&mut cursor) {
        if item.kind() == "string" {
            out.push(string_literal_value(item, bytes));
        }
    }
}

fn string_literal_value(node: Node, bytes: &[u8]) -> String {
    let mut cursor = node.walk();
    for c in node.named_children(&mut cursor) {
        if c.kind() == "string_content" {
            if let Ok(t) = c.utf8_text(bytes) {
                return t.to_string();
            }
        }
    }
    unquote(node.utf8_text(bytes).unwrap_or("")).to_string()
}

fn mk_sym(name: &str, kind: SymbolKind, node: &Node) -> ExportedSymbol {
    ExportedSymbol {
        name: name.to_string(),
        kind,
        flags: SymbolFlag::NONE,
        parent_qualifier: None,
        source_path: None,
        line: node.start_position().row as u32,
        col: node.start_position().column as u32,
    }
}
